ocaml/parsing/parser.mly

```diff
@@ -4263,6 +4298,30 @@ strict_function_or_labeled_tuple_type:
    | { Mode.empty }
    | mode_expr_legacy_nonempty {$1}
 ;
+
+/* New mode annotation, followed by AT or ATAT and thus enjoys its own namespace */
+%inline mode:
+  | LIDENT { Mode.Const.mk $1 (make_loc $sloc) }
+;
+
+%inline mode_expr:
+  | mode+ {
+    mkloc $1 (make_loc $sloc)
+  }
+;
+
+%inline at_mode_expr_nonempty:
+  | AT mode_expr {$2}
+;
+
+%inline atat_mode_expr_nonempty:
+  | ATAT mode_expr {$2}
+
+%inline atat_mode_expr:
+  | { Mode.empty }
+  | atat_mode_expr_nonempty {$1}
+;
```

Comment:
Nit: I would find it more understandable to have `atat_mode_expr`, `at_mode_expr`, and `optional_atat_mode_expr`. This makes it easier for me to understand which clients expect `@@` and `@` to necessarily show up in the syntax.